### PR TITLE
Variable dimension FFT

### PR DESCRIPTION
**Pull request by streeve**
Add 2D for FFT base and heFFTe

Part of #340

### CODE REVIEW

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 1]
```diff
@@ -407,24 +488,15 @@ class HeffteFastFourierTransform
...
-            "fft_copy_x_to_cufft_work",
-            createExecutionPolicy( own_space, exec_space() ),
-            KOKKOS_LAMBDA( const int i, const int j, const int k ) {
-                auto iw = i - own_space.min( Dim::I );
-                auto jw = j - own_space.min( Dim::J );
-                auto kw = k - own_space.min( Dim::K );
-                work_view( iw, jw, kw, 0 ) = x_view( i, j, k, 0 );
-                work_view( iw, jw, kw, 1 ) = x_view( i, j, k, 1 );
-            } );
+        this->copyToLocal( own_space, local_view, localghost_view );
```
Comment (sfogerty) [inline cajita/src/Cajita_FastFourierTransform.hpp:499]: Much nicer 👍

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 2]
```diff
@@ -329,14 +400,24 @@ class HeffteFastFourierTransform
...
     using heffte_backend_type =
         typename Impl::HeffteBackendTraits<exec_space,
                                            backend_type>::backend_type;
+    using heffte_fft_type = std::conditional_t<
+        3 == num_space_dim, heffte::fft3d<heffte_backend_type>,
+        std::conditional_t<2 == num_space_dim,
+                           heffte::fft2d<heffte_backend_type>, void>>;
+    using heffte_box_type = std::conditional_t<
+        3 == num_space_dim, heffte::box3d,
+        std::conditional_t<2 == num_space_dim, heffte::box2d, void>>;
```
Comment (sfogerty) [inline cajita/src/Cajita_FastFourierTransform.hpp:415]: These may be unnecessary. heFFTe's `fft2d` and `box2d` are just aliases for `fft3d` and `box3d`. What I understand is that heFFTe should be able to detect geometry and act appropriately.

Comment (streeve) [reply cajita/src/Cajita_FastFourierTransform.hpp]: Good point - I simplified and commented that 3d can also handle 2d

**Review by sfogerty — APPROVED** [review]

**Review by sslattery — APPROVED** [review]
very nice

**Pull request merged**